pulp_ansible/app/galaxy/v3/views.py

```diff
@@ -126,68 +157,92 @@ class CollectionViewSet(
         """
         Returns a Collections queryset for specified distribution.
         """
-        repo_version = self.get_repository_version(self.kwargs["path"])
-        deprecated_query = AnsibleCollectionDeprecated.objects.filter(
-            collection=OuterRef("pk"), repository_version=repo_version
+        repo_version = self._repository_version
+        return Collection.objects.filter(
+            versions__in=repo_version.content,
         )
-        collections = Collection.objects.filter(versions__in=repo_version.content).distinct()
-        collections = collections.annotate(deprecated=Exists(deprecated_query))
-
-        versions_qs = CollectionVersion.objects.filter(pk__in=repo_version.content).values_list(
-            "collection_id",
-            "namespace",
-            "name",
-            "version",
-            "pulp_created",
+
+    def append_context(self, queryset):
+        """Appending collection data to context."""
+        repo_version = self._repository_version
+        collections_qs = queryset.annotate(available_versions=ArrayAgg("versions__version"))
+        versions_context = {}
+        for collection_id, available_versions in collections_qs.values_list(
+            "pk", "available_versions"
+        ):
+            versions_context[collection_id] = available_versions
+
+        self.available_versions_context = versions_context  # needed by get__serializer_context
+        self._deprecation
+
+        collections = Collection.objects.filter(
+            pk__in=versions_context.keys(),
+            versions__version_memberships__repository=repo_version.repository,
+        ).annotate(
+            repo_version_added_at=Window(
+                expression=FirstValue(
+                    "versions__version_memberships__version_added__pulp_last_updated"
+                ),
+                partition_by=[F("versions__collection_id")],
+                order_by=F("versions__version_memberships__version_added__pulp_last_updated").desc(
+                    nulls_last=True
+                ),
+            ),
+            repo_version_removed_at=Window(
+                expression=FirstValue(
+                    "versions__version_memberships__version_removed__pulp_last_updated"
+                ),
+                partition_by=[F("versions__collection_id")],
+                order_by=F(
+                    "versions__version_memberships__version_removed__pulp_last_updated"
+                ).desc(nulls_last=True),
+            ),
         )
 
-        highest_versions = defaultdict(
-            lambda: CollectionTuple(None, None, semantic_version.Version("0.0.0"), None)
+        return collections.distinct("versions__collection_id").only(
+            "pulp_created", "name", "namespace"
         )
-        lowest_versions = defaultdict(
-            lambda: CollectionTuple(None, None, semantic_version.Version("100000000000.0.0"), None)
+
+    def paginate_queryset(self, queryset):
```

Comment:
Why do we need this custom paginator?

Reply:
https://github.com/encode/django-rest-framework/blob/master/rest_framework/pagination.py#L363-L391

Reply:
for avoiding this query for all data:
```sql
SELECT COUNT(*)
  FROM (
        SELECT DISTINCT
            ON ("ansible_collectionversion"."collection_id") "ansible_collection"."pulp_id" AS Col1,
               "ansible_collection"."pulp_created" AS Col2,
               "ansible_collection"."namespace" AS Col3,
               "ansible_collection"."name" AS Col4,
               FIRST_VALUE("core_repositoryversion"."pulp_last_updated") OVER (PARTITION BY "ansible_collectionversion"."collection_id" ORDER BY "core_repositoryversion"."pulp_last_updated" DESC NULLS LAST) AS "repo_version_added_at",
               FIRST_VALUE(T8."pulp_last_updated") OVER (PARTITION BY "ansible_collectionversion"."collection_id" ORDER BY T8."pulp_last_updated" DESC NULLS LAST) AS "repo_version_removed_at"
          FROM "ansible_collection"
         INNER JOIN "ansible_collectionversion"
            ON ("ansible_collection"."pulp_id" = "ansible_collectionversion"."collection_id")
         INNER JOIN "core_content"
            ON ("ansible_collectionversion"."content_ptr_id" = "core_content"."pulp_id")
         INNER JOIN "core_repositorycontent"
            ON ("core_content"."pulp_id" = "core_repositorycontent"."content_id")
         INNER JOIN "core_repositoryversion"
            ON ("core_repositorycontent"."version_added_id" = "core_repositoryversion"."pulp_id")
          LEFT OUTER JOIN "core_repositoryversion" T8
            ON ("core_repositorycontent"."version_removed_id" = T8."pulp_id")
         WHERE ("ansible_collection"."pulp_id" IN ('9c4e0be5-47bb-48a1-a951-aa2e43590488'::uuid) AND "core_repositorycontent"."repository_id" = 'ba5f0b1f-2012-461d-b580-98541236e1cd'::uuid)
       ) subquery
```

Reply:
With a custom pagination we have:
```sql
SELECT COUNT(*)
  FROM (
        SELECT DISTINCT
            ON ("ansible_collectionversion"."collection_id") "ansible_collection"."pulp_id" AS Col1
          FROM "ansible_collection"
         INNER JOIN "ansible_collectionversion"
            ON ("ansible_collection"."pulp_id" = "ansible_collectionversion"."collection_id")
         WHERE "ansible_collectionversion"."content_ptr_id" IN (
                SELECT V0."pulp_id"
                  FROM "core_content" V0
                 INNER JOIN "core_repositorycontent" V1
                    ON (V0."pulp_id" = V1."content_id")
                 WHERE V1."pulp_id" IN (
                        SELECT U0."pulp_id"
                          FROM "core_repositorycontent" U0
                         INNER JOIN "core_repositoryversion" U2
                            ON (U0."version_added_id" = U2."pulp_id")
                          LEFT OUTER JOIN "core_repositoryversion" U3
                            ON (U0."version_removed_id" = U3."pulp_id")
                         WHERE (U0."repository_id" = 'ba5f0b1f-2012-461d-b580-98541236e1cd'::uuid AND U2."number" <= 1 AND NOT (U3."number" <= 1 AND U3."number" IS NOT NULL))
                       )
               )
       ) subquery
```